frame/beefy/src/lib.rs

```diff
@@ -20,46 +20,84 @@
 use codec::{Encode, MaxEncodedLen};
 
 use frame_support::{
+	dispatch::{DispatchResultWithPostInfo, Pays},
 	log,
-	traits::{Get, OneSessionHandler},
+	pallet_prelude::*,
+	traits::{Get, KeyOwnerProofSystem, OneSessionHandler},
+	weights::Weight,
 	BoundedSlice, BoundedVec, Parameter,
 };
-
+use frame_system::{
+	ensure_none, ensure_signed,
+	pallet_prelude::{BlockNumberFor, OriginFor},
+};
 use sp_runtime::{
 	generic::DigestItem,
 	traits::{IsMember, Member},
-	RuntimeAppPublic,
+	KeyTypeId, RuntimeAppPublic,
 };
+use sp_session::{GetSessionNumber, GetValidatorCount};
+use sp_staking::SessionIndex;
 use sp_std::prelude::*;
 
 use beefy_primitives::{
-	AuthorityIndex, ConsensusLog, OnNewValidatorSet, ValidatorSet, BEEFY_ENGINE_ID,
-	GENESIS_AUTHORITY_SET_ID,
+	AuthorityIndex, BeefyAuthorityId, ConsensusLog, EquivocationProof, OnNewValidatorSet,
+	ValidatorSet, BEEFY_ENGINE_ID, GENESIS_AUTHORITY_SET_ID,
 };
 
+mod default_weights;
+mod equivocation;
 #[cfg(test)]
 mod mock;
-
 #[cfg(test)]
 mod tests;
 
+pub use crate::equivocation::{
+	BeefyEquivocationOffence, BeefyOffence, BeefyTimeSlot, EquivocationHandler, HandleEquivocation,
+};
 pub use pallet::*;
 
+const LOG_TARGET: &str = "runtime::beefy";
+
 #[frame_support::pallet]
 pub mod pallet {
 	use super::*;
-	use frame_support::pallet_prelude::*;
 	use frame_system::pallet_prelude::BlockNumberFor;
 
 	#[pallet::config]
 	pub trait Config: frame_system::Config {
 		/// Authority identifier type
 		type BeefyId: Member
 			+ Parameter
-			+ RuntimeAppPublic
+			// todo: use custom signature hashing type
```

Comment:
What is meant here? Is it to be able to allow defining the hash type on the pallet, so that it could be passed along here?

Reply:
yes, exactly. currently we define `BeefyAuthorityId<sp_runtime::traits::Keccak256>` hardcoding `Keccak256` as the signature hashing type - we might want custom hashing, but don't know for sure yet, leaving this `todo`

Reply:
opened https://github.com/paritytech/substrate/issues/13403 to track this potential enhancement